### ISSUE TITLE
[Bug]: PWA Icon Doesn't Display When Using Reverse Proxy

### ISSUE THREAD

**Issue by louisefindlay23** [coder/code-server#5248]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS/Web Information

- Web Browser: Safari
- Local OS: iPadOS 15.5
- Remote OS: Libreelec (Linux Server Docker)
- Remote Architecture: x86_64
- `code-server --version`: v4.4.0

### Steps to Reproduce

1. Reverse Proxy code-server
2. Open code-server via the proxied subdomain on an iPad using Safari
3. Tap the Share Button
4. Press Add to Home Screen

### Expected

The home screen icon should resemble the favicons in the favicon folder

### Actual

See the home screen icon which is a screenshot of the VS Code workbench (current page view)

### Does this issue happen in VS Code?

- [X] I cannot reproduce this in VS Code.

### Are you accessing code-server over HTTPS?

- [X] I am using HTTPS.

### Notes

Tried to add favicon routes using Nginx but that didn't work either.

**Comment by jsjoeio** [comment]
I don't know how we would fix this?  🤔 Do you?

**Comment by louisefindlay23** [comment]
> Does this happen with any site that is reverse proxied or just code-server?

No, only some. Organizr, Pihole and Calibre-Web display the favicon correctly when adding to the homescreen. Home Assistant and code-server don't.

**Comment by jsjoeio** [comment]
Interesting...did you file an issue with Home Assistant too? Curious what they said

**Comment by louisefindlay23** [comment]
No, I haven't but have just done so. Will let you know what they say.

**Comment by jsjoeio** [comment]
Do you mind linking the issue here?

**Comment by louisefindlay23** [comment]
Sure: https://github.com/home-assistant/frontend/issues/12931

**Comment by icui** [comment]
Have you signed with services like Let's Encrypt? I think iOS will not load the icon if you manually trust the certificate (like the default certificate).

**Comment by louisefindlay23** [comment]
> Have you signed with services like Let's Encrypt? I think iOS will not load the icon if you manually trust the certificate (like the default certificate).

I have my reverse proxy signed with Let's Encrypt but code-server is HTTP.

**Comment by bilogic** [comment]
This is likely an issue with Safari and PWA standards. My android 11 app and home screen shows the same icon as per favicon in Chrome browser.

Is there anyway to customize the name and favicon of the PWA? (or should I open a new feature issue?)

**Comment by jsjoeio** [comment]
@bilogic no there isn't at the moment!

**Comment by code-asher** [comment]
I am going through old issues, is this still a problem?  I have no iPad to test, but maybe there is some PWA icon path we are missing in the manifest or something.

**Comment by bilogic** [comment]
Hmm, if manifest is the issue, then nothing to do with reverse proxy. Should modify the title to reflect that if confirmed.

**Comment by code-asher** [comment]
This is interesting, if I put code-server behind a reverse proxy and a base path and check the manifest in Chromium, it tells me the icons failed to load.  And yet if I open the icons directly in my browser, the images load just fine.  Very confusing.

If I access code-server directly on localhost, I get no errors.

**Comment by bilogic** [comment]
I'm on reverse proxy too. Should be due to some missing configuration. 
This is my full config on Apache and I have forgotten how complicated it was:
```
Header always set Strict-Transport-Security "max-age=63072000; includeSubdomains;"
RewriteEngine On
RewriteCond %{HTTP:Upgrade} =websocket [NC]
RewriteRule /(.*) ws://127.0.0.1:8001/$1 [P,L]
RewriteCond %{HTTP:Upgrade} !=websocket [NC]
RewriteRule /(.*) http://127.0.0.1:8001/$1 [P,L]
ProxyPreserveHost on
ProxyRequests off
ProxyPass / http://127.0.0.1:8001/ nocanon
ProxyPassReverse / http//127.0.0.1:8001
```

**Comment by bilogic** [comment]
![image](https://github.com/coder/code-server/assets/946010/928dd232-c467-43c1-90d8-56e09dc9ff6a)

1. The cause might be the `Start URL`, github has a `/` instead of `.`
2. Should we be concerned about the Errors and warnings?

**Comment by code-asher** [comment]
Oh interesting.  Maybe it only happens if you serve at some base path (like `domain.tld/code-server/`) and not at the root? 

I thought a relative URL for the start URL was allowed, but maybe it is causing issues.

I am not sure about the other warnings...from what I have read, you can still install the PWA, but it just will not use the richer fancier-looking dialog.

**Comment by bilogic** [comment]
@code-asher are you or @louisefindlay23 serving on a path?

I think the warning is saying `code-server` has to add a screenshot or something, otherwise richer UI is unavailable.

**Comment by code-asher** [comment]
> @code-asher are you or @louisefindlay23 serving on a path?

Yeah I am, not sure about @louisefindlay23 though.

![screenshot](https://github.com/coder/code-server/assets/45609798/d63e68fe-8101-4ee8-bbf2-2697315232f2)

![screenshot](https://github.com/coder/code-server/assets/45609798/652809d6-30a0-4ad3-ac01-a6e374f18532)

> I think the warning is saying code-server has to add a screenshot or something, otherwise richer UI is unavailable.

That sounds right to me.  It would be nice to support the richer UI, but this is probably not causing any bugs at least.

**Comment by bilogic** [comment]
Can also show your manifest.json?

**Comment by code-asher** [comment]
Mine is the same other than the name.  Makes sense though since it is all relative URLs:

```
{
  "name": "code-server",
  "short_name": "code-server",
  "start_url": ".",
  "display": "fullscreen",
  "display_override": [
    "window-controls-overlay"
  ],
  "description": "Run Code on a remote server.",
  "icons": [
    {
      "src": "./_static/src/browser/media/pwa-icon-192.png",
      "type": "image/png",
      "sizes": "192x192"
    },
    {
      "src": "./_static/src/browser/media/pwa-icon-512.png",
      "type": "image/png",
      "sizes": "512x512"
    }
  ]
}
```

Nothing in the console stands out to me unfortunately:

```
Navigated to https://dev.coder.com/@code-asher/dev27.dev/apps/code-server/?folder=%2Fhome%2Fcoder%2Fcoder
INFO Resolving connection token (dev.coder.com)...
INFO Resolved connection token (dev.coder.com) after 4 ms
INFO Creating a socket (renderer-Management-143bf21f-26c0-4551-9a05-d5d04c691522)...
INFO Creating a socket (renderer-Management-143bf21f-26c0-4551-9a05-d5d04c691522) was successful after 509 ms.
GET https://dev.coder.com/@code-asher/dev27.dev/apps/code-server/stable-9a28bc29dbddb6886dfe03dc1c31320249a901ce/static/out/vsda_bg.wasm 404 (Not Found)
GET https://dev.coder.com/@code-asher/dev27.dev/apps/code-server/stable-9a28bc29dbddb6886dfe03dc1c31320249a901ce/static/out/vsda.js net::ERR_ABORTED 404 (Not Found)
Refused to execute script from 'https://dev.coder.com/@code-asher/dev27.dev/apps/code-server/stable-9a28bc29dbddb6886dfe03dc1c31320249a901ce/static/out/vsda.js' because its MIME type ('text/plain') is not executable, and strict MIME type checking is enabled.
The web worker extension host is started in a same-origin iframe!
INFO Creating a socket (renderer-ExtensionHost-0d2f58fc-1fe5-4bf1-9a7b-78f5eaaf0aef)...
INFO [Service Worker] registered
An iframe which has both allow-scripts and allow-same-origin for its sandbox attribute can escape its sandboxing.
INFO Creating a socket (renderer-ExtensionHost-0d2f58fc-1fe5-4bf1-9a7b-78f5eaaf0aef) was successful after 501 ms.
INFO Updating additional builtin extensions cache
```

The weird thing to me is, Chromium must be resolving the icon URL correctly, because the URL printed in the error is the right one.  Yet it still cannot load it for some reason.  Unless the URL it prints and the URL it actually tries to load are different... :thinking: 

Firefox loads the icon without any issues, so maybe this is a Webkit bug.

**Comment by code-asher** [comment]
Oh yeah I checked the network tab too, there are no failed requests for those icons.  :confused:

**Comment by bilogic** [comment]
> Oh yeah I checked the network tab too, there are no failed requests for those icons. 😕 There is a successful request for `pwa-icon-192.png` though.

You might have to go into application, unregister the service work, have dev console opened and load your `code-server` again

**Comment by louisefindlay23** [comment]
> > @code-asher are you or @louisefindlay23 serving on a path?
> 
> Yeah I am, not sure about @louisefindlay23 though.
> 
> ![screenshot](https://private-user-images.githubusercontent.com/45609798/327610102-d63e68fe-8101-4ee8-bbf2-2697315232f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ3NTA4OTQsIm5iZiI6MTcxNDc1MDU5NCwicGF0aCI6Ii80NTYwOTc5OC8zMjc2MTAxMDItZDYzZTY4ZmUtODEwMS00ZWU4LWJiZjItMjY5NzMxNTIzMmYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAzVDE1MzYzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjNjBiODg5MjMzYmZkYTVmMjNjM2E4MjU3ZTJhOWUyMGNhZmYyMzYxNjA2ZmVhMGI4ZTkwNDcyNmE2MDBmYzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Jl8QBfm2o2zhlZfz4HvjD83L0qg_6Fo4Z3lTNek41Dw)
> 
> ![screenshot](https://private-user-images.githubusercontent.com/45609798/327610011-652809d6-30a0-4ad3-ac01-a6e374f18532.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ3NTA4OTQsIm5iZiI6MTcxNDc1MDU5NCwicGF0aCI6Ii80NTYwOTc5OC8zMjc2MTAwMTEtNjUyODA5ZDYtMzBhMC00YWQzLWFjMDEtYTZlMzc0ZjE4NTMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAzVDE1MzYzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0OTNlODBjYjQ5YzllMTkwNzEwZDNlNzI2MDY5ZjgzMGU4MDhjMWQyNzYyODhkNDAzZGNmM2ViMjYzZjc5ZWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TbQVfofhxA9fPXS92PFqRM4ChLceY5Xf_K5CMih8NKo)
> 
> > I think the warning is saying code-server has to add a screenshot or something, otherwise richer UI is unavailable.
> 
> That sounds right to me. It would be nice to support the richer UI, but this is probably not causing any bugs at least.

Yes, I'm serving on a path.

**Comment by bilogic** [comment]
How about we try serving on root first? To possibly narrow down the culprit: Browser? code-server Manifest? Or VSCode?

Also side track, @jsjoeio would you so kind as to help review this https://github.com/microsoft/vscode/pull/207721? It needs some community reviews from existing contributors

**Comment by bilogic** [comment]
@code-asher

The problem description here seems to match https://stackoverflow.com/questions/59520750/manifest-json-failing-to-load-icon-files

But I don't see code-server sending the `Cross-Origin-Resource-Policy` header causing it

**Comment by code-asher** [comment]
> How about we try serving on root first

Yeah, no errors if I serve from root or localhost.

Since it only happens on Chromium it does seem like a Webkit bug but who knows.  :shrug:

> But I don't see code-server sending the Cross-Origin-Resource-Policy header causing it

Hmm yeah I am not seeing that either.

**Comment by bilogic** [comment]
@code-asher don't mind are you currently a vscode contributor? I need a review here https://github.com/microsoft/vscode/pull/207721 and could use some votes here https://github.com/microsoft/vscode/issues/201452

I think an easier way out that everyone can chip in is to find another PWA that works on a path and can show icons correctly. Then we can have an apple vs orange comparison to quickly narrow things down.

**Comment by code-asher** [comment]
I am not a VS Code contributor, unfortunately.  :cry:  I added my vote though.

> find another PWA that works on a path

Yeah that would be awesome.  I will keep my eye out.
